File: demo4.py
def lause_toisinpain(lause):
  sana = ""
  kaannetty_lause = ""
  for x in lause:
    if x != " ":
      sana += x
    elif x == " " and len(sana) < 1:
      kaannetty_lause += " "
    elif len(sana) > 0 :
      kaannetty_lause = sana + " " + kaannetty_lause
      sana = ""

  if len(sana) > 0:
    kaannetty_lause = sana + " " + kaannetty_lause
  return kaannetty_lause
lause = " algoritmien opiskelu   on kivaa"
print(lause_toisinpain(lause))

File: Demo4.java
public class Demo4 {
  public static String[] haeSanat(String lause) {
    String[] arr = new String[100];
    String sana = "";
    int j = -1;
    for (int i = 0; i < lause.length(); i++) {
      char x = lause.charAt(i);
      if (x != ' ') {
        sana += x;
      } else if (sana.length() > 0) {
        j++;
        arr[j] = sana;
        sana = "";
      }
    }
    return arr;
  }

  public static void kaanna(String lause, String[] sanat) {
    String str = "";
    int sanaIndex = 0;
    while (sanat[sanaIndex + 1] != null)
      sanaIndex++;

    for (int i = 0; i < lause.length(); i++) {
      char x = lause.charAt(i);
      if (x == ' ') {
        str += x;
      } else {
        str += sanat[sanaIndex];
        sanaIndex--;
        while (lause.charAt(i + 1) != ' ') {
          i++;          
        }
      }
    }
    System.out.println(str);
  }

  public static void main(String[] args) {
    String lause = "  algoritmien opiskelu on   kivaa  ";
    // String lause = "  a  ";
    // String lause = "    ";
    // String lause = "algoritmien opiskelu on kivaa";

    String[] sanat = haeSanat(lause);
    kaanna(lause, sanat);
  }
}